src/Features/Core/Portable/CodeRefactorings/ConvertNumericLiteral/AbstractConvertNumericLiteralCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,166 @@
+                return;
+            }
+
+            if (!IsIntegral(symbol.SpecialType))
+            {
+                return;
+            }
+
+            var valueOpt = semanticModel.GetConstantValue(syntaxNode);
+            if (!valueOpt.HasValue)
+            {
+                return;
+            }
+
+            var value = IntegerUtilities.ToInt64(valueOpt.Value);
+            var numericText = numericToken.ToString();
+            var (hexPrefix, binaryPrefix) = GetNumericLiteralPrefixes();
+            var (prefix, number, suffix) = GetNumericLiteralParts(numericText, hexPrefix, binaryPrefix);
+            var kind = string.IsNullOrEmpty(prefix) ? NumericKind.Decimal
+                : prefix.Equals(hexPrefix, StringComparison.OrdinalIgnoreCase) ? NumericKind.Hexadecimal
+                : prefix.Equals(binaryPrefix, StringComparison.OrdinalIgnoreCase) ? NumericKind.Binary
+                : NumericKind.Unknown;
+
+            if (kind == NumericKind.Unknown)
+            {
+                return;
+            }
+
+            if (kind != NumericKind.Decimal)
+            {
+                RegisterRefactoringWithResult(value.ToString(), FeaturesResources.Convert_to_decimal);
+            }
+
+            if (kind != NumericKind.Binary)
+            {
+                RegisterRefactoringWithResult(binaryPrefix + Convert.ToString(value, 2), FeaturesResources.Convert_to_binary);
+            }
+
+            if (kind != NumericKind.Hexadecimal)
+            {
+                RegisterRefactoringWithResult(hexPrefix + value.ToString("X"), FeaturesResources.Convert_to_hex);
+            }
+
+            const string DigitSeparator = "_";
+            if (numericText.Contains(DigitSeparator))
+            {
+                RegisterRefactoringWithResult(prefix + number.Replace(DigitSeparator, string.Empty), FeaturesResources.Remove_separators);
+            }
+            else
+            {
+                switch (kind)
+                {
+                    case NumericKind.Decimal:
+                        RegisterRefactoringWithResult(AddSeparators(number, 3), FeaturesResources.Separate_thousands);
+                        break;
+
+                    case NumericKind.Hexadecimal:
+                        RegisterRefactoringWithResult(hexPrefix + AddSeparators(number, 4), FeaturesResources.Separate_words);
+                        break;
+
+                    case NumericKind.Binary:
+                        RegisterRefactoringWithResult(binaryPrefix + AddSeparators(number, 4), FeaturesResources.Separate_nibbles);
+                        break;
+                }
+            }
+
+            void RegisterRefactoringWithResult(string text, string title)
+            {
+                var generator = SyntaxGenerator.GetGenerator(document);
+                var updatedToken = generator.NumericLiteralToken(text + suffix, (ulong)value);
+                var updatedRoot = root.ReplaceToken(numericToken, updatedToken);
```

Comment:
@CyrusNajmabadi Quick question: would it better to move these locals to the lambda passed to RegisterRefactoring below? I want to know specifically when that lambda is executed and whether the result is cached or not. Thanks!